src/transformers/modeling_utils.py

```diff
@@ -465,6 +465,74 @@ def __init__(self, config: PretrainedConfig, *inputs, **kwargs):
         self.config = config
         self.name_or_path = config.name_or_path
 
+    @classmethod
+    def _from_config(cls, config, **kwargs):
+        """
+        All context managers that the model should be init'ed under go here
+        """
+
+        # override default dtype if needed
+        dtype_orig = cls._set_default_dtype(config.torch_dtype)
+
+        if is_deepspeed_zero3_enabled():
+            import deepspeed
+
+            logger.info("Detected DeepSpeed ZeRO-3: activating zero.init() for this model")
+            # this immediately partitions the model across all gpus, to avoid the overhead in time
+            # and memory copying it on CPU or each GPU first
+            with deepspeed.zero.Init(config=deepspeed_config()):
+                model = cls(config, **kwargs)
+        else:
+            model = cls(config, **kwargs)
+
+        # restore default dtype if it was modified
+        if dtype_orig is not None:
+            torch.set_default_dtype(dtype_orig)
+
+        return model
+
+    @classmethod
+    def _set_default_dtype(cls, config_dtype_str, weight_dtype=None) -> torch.dtype:
+        """
+        If a floating ``dtype`` is passed via the model config or one of its weights ``torch.set_default_dtype`` is
+        used to change the global dtype. Which is needed when wanting to instantiate the model under specific dtype.
+
+        Args:
+            config_dtype_str (:obj:`str`):
+                value of ``config.torch_dtype``
+            weight_dtype (:obj:`torch.dtype`, `optional`, defaults to :obj:`None`):
+                ``dtype`` of one of the pretrained weights
+
+        Returns:
+            :obj:`torch.dtype`: the original ``dtype`` that can be used to restore ``torch.set_default_dtype(dtype)``
+            if it was modified. If it wasn't, returns :obj:`None`.
+
+        Note ``set_default_dtype`` currently only works with floating-point types and asserts if for example,
+        ``torch.int64`` is passed. So if a non-float ``dtype`` is passed we don't do anything other than logging a
+        warning that the non-float dtype was ignored.
+        """
+        dtype_str = config_dtype_str
+        if dtype_str is None and weight_dtype is not None:
+            dtype_str = str(weight_dtype).split(".")[1]
+
+        if dtype_str is None:
+            return None
+
+        if not isinstance(dtype_str, str):
+            raise ValueError(f"dtype is expected to be the string attribute name of ``torch``, got {config_dtype_str}")
+
+        dtype = getattr(torch, dtype_str)
+        if not dtype.is_floating_point:
+            logger.warning(
+                f"instantiating {cls.__name__} model under ``torch.get_default_dtype()``, since {dtype} is not a floating point dtype"
```

Comment:
```suggestion
                f"instantiating {cls.__name__} model under {torch.get_default_dtype()}, since {dtype} is not a floating point dtype"
```
Maybe show the value instead of the function name?

Reply:
Great idea!